File: solutions/c/dnd-character/1/dnd_character.c
#include <time.h>
#include <stddef.h>
#include <math.h>
#include <stdlib.h>  
#include "dnd_character.h"

int ability(void) {
    time(NULL);
    int data[4];
    int res = 0;
    for (int i = 0; i < 4; ++i) {
        data[i] = 1 + rand() % 6;
        res += data[i];
    }
    int min = 7;
    for (int i = 0; i < 4; ++i) {
        if (min > data[i])
            min = data[i];
    }
    return res - min;
}

int modifier(int score) {
    return floor((score - 10.0) / 2.0);
}

dnd_character_t make_dnd_character(void) {
    dnd_character_t res;
    res.strength = ability();
    res.dexterity = ability();
    res.constitution = ability();
    res.intelligence = ability();
    res.wisdom = ability();
    res.charisma = ability();
    res.hitpoints = 10 + modifier(res.constitution);
    return res;
}

File: solutions/go/dnd-character/1/dnd_character.go
package dndcharacter

import (
    "math"
    "math/rand"
    "time"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return int(math.Floor((float64(score) - 10.0) / 2.0))
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	rand.Seed(time.Now().UnixNano())
    data := []int{}
    res := 0
    for i := 0; i < 4; i++ {
        data = append(data, 1 + rand.Intn(6))
        res += data[i]
    }
    min := 7
    for i := 0; i < 4; i++ {
    	if min > data[i] {
            min = data[i]
        }
    }
    return res - min
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	var ch Character
    ch.Strength = Ability()
	ch.Dexterity = Ability()
	ch.Constitution = Ability()
	ch.Intelligence = Ability()
	ch.Wisdom = Ability()
	ch.Charisma = Ability()
	ch.Hitpoints = 10 + Modifier(ch.Constitution)
    return ch
}
